weblab/entities/tests/test_views.py

```diff
@@ -2054,3 +2054,351 @@ def test_nonexistent_entity_redirects_anonymous_to_login(self, client, helpers,
+        helpers.add_permission(logged_in_user, 'create_experiment', Experiment)
+        model = recipes.model.make(author=logged_in_user)
+        commit1 = helpers.add_version(model, visibility='public')
+        commit2 = helpers.add_version(model, visibility='public')
+        model.add_tag('v1', commit2.hexsha)
+        protocol = recipes.protocol.make(author=logged_in_user)
+        commit_protocol = helpers.add_version(protocol, visibility='public')
+        response = client.get('/entities/protocols/%d/runexperiments/' % protocol.pk)
+        assert response.status_code == 200
+        assert response.context['object_list'] == [{'id': model.pk,
+                                                    'name': 'mymodel1',
+                                                    'versions': [{'commit': commit2, 'tags': ['v1'], 'latest': True},
+                                                                 {'commit': commit1, 'tags': [], 'latest': False}]},
+                                                   ]
+        # Test post returns correct response
+        data = {'runexperimentlist[]': ['%d:%s' % (model.pk, commit1.hexsha), '%d:%s' % (model.pk, commit2.hexsha)],
+                'entity.repo.latest_commit.hexsha': commit_protocol.hexsha,
+                'rerun_expts': 'on'}
+        response = client.post('/entities/protocols/%d/runexperiments/' % protocol.pk, data=data)
+        assert response.status_code == 302
+        assert response.url == '/entities/protocols/%d/versions/latest' % protocol.pk
+
+        # Test that planned experiments have been added correctly
+        expected_model_versions = set([
+            (model, commit2.hexsha),
+            (model, commit1.hexsha)
+        ])
+        assert PlannedExperiment.objects.count() == 2
+        for planned_experiment in PlannedExperiment.objects.all():
+            assert planned_experiment.protocol == protocol
+            assert planned_experiment.protocol_version == commit_protocol.hexsha
+            assert (planned_experiment.model, planned_experiment.model_version) in expected_model_versions
+
+    def test_view_run_experiment_protocol_post_exclude_existing(self, client, helpers, logged_in_user):
+        helpers.add_permission(logged_in_user, 'create_experiment', Experiment)
+        model = recipes.model.make(author=logged_in_user)
+        commit1 = helpers.add_version(model, visibility='public')
+        commit2 = helpers.add_version(model, visibility='public')
+        model.add_tag('v1', commit2.hexsha)
+        protocol = recipes.protocol.make(author=logged_in_user)
+        commit_protocol = helpers.add_version(protocol, visibility='public')
+
+        recipes.experiment_version.make(
+            status='SUCCESS',
+            experiment__model=model,
+            experiment__model_version=commit1.hexsha,
+            experiment__protocol=protocol,
+            experiment__protocol_version=commit_protocol.hexsha)
+
```

Comment:
To test behaviour with an experiment that has no versions, add:
```python
        recipes.experiment.make(
            model=model,
            model_version=commit2.hexsha,
            protocol=protocol,
            protocol_version=commit_protocol.hexsha)
```

Reply:
added